server/http/filters/filters.go

```diff
@@ -93,14 +104,96 @@ func RequestID(next http.Handler) http.Handler {
 	})
 }
 
+// instrumentedResponseWriter wraps http.ResponseWriter, recording info needed for
+// Prometheus metrics.
+type instrumentedResponseWriter struct {
+	http.ResponseWriter
+
+	statusCode        int
+	responseSizeBytes int
+}
+
+func (w instrumentedResponseWriter) Header() http.Header {
+	return w.ResponseWriter.Header()
+}
+func (w instrumentedResponseWriter) Write(bytes []byte) (int, error) {
+	w.responseSizeBytes += len(bytes)
+	return w.ResponseWriter.Write(bytes)
+}
+func (w instrumentedResponseWriter) WriteHeader(statusCode int) {
+	w.statusCode = statusCode
+	w.ResponseWriter.WriteHeader(statusCode)
+}
+func (w instrumentedResponseWriter) Hijack() (net.Conn, *bufio.ReadWriter, error) {
+	return w.ResponseWriter.(http.Hijacker).Hijack()
+}
+func (w instrumentedResponseWriter) Flush() {
+	w.ResponseWriter.(http.Flusher).Flush()
+}
+
 func LogRequest(next http.Handler) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		start := time.Now()
-		next.ServeHTTP(w, r)
-		log.LogHTTPRequest(r.Context(), r.URL.Path, time.Now().Sub(start), nil)
+		method := getMethod(r)
+		route := getRoute(r)
+		recordRequestMetrics(route, method)
+		ow := instrumentedResponseWriter{
+			ResponseWriter: w,
+		}
+		next.ServeHTTP(ow, r)
+		duration := time.Since(start)
+		log.LogHTTPRequest(r.Context(), r.URL.Path, duration, nil)
+		recordResponseMetrics(route, method, ow, duration)
 	})
 }
 
+func getRoute(r *http.Request) string {
+	path := r.URL.Path
+
+	// TODO(bduffany): migrate to a routing solution that doesn't require
+	// updating this function when we add new HTTP routes.
+
+	// Strip prefixes on large static file directories to avoid
+	// creating new metrics series for every static file that we serve.
+	if strings.HasPrefix(path, "/images/") {
+		return "/images/:path"
+	}
+	if strings.HasPrefix(path, "/favicon/") {
+		return "/favicon/:path"
+	}
+
+	// Replace path parameters to avoid creating a series for
+	// every possible parameter value.
+
+	// Currently we only use UUID v4 path params so this
+	// find-and-replace is good enough for now.
+	return uuidV4Regexp.ReplaceAllLiteralString(path, ":id")
+}
+
+func getMethod(r *http.Request) string {
+	if r.Method == "" {
+		return "GET"
+	}
+	return r.Method
+}
+
+func recordRequestMetrics(route, method string) {
+	metrics.HTTPRequestCount.With(prometheus.Labels{
+		metrics.HTTPRouteLabel:  route,
+		metrics.HTTPMethodLabel: method,
+	}).Inc()
+}
+
+func recordResponseMetrics(route, method string, w instrumentedResponseWriter, duration time.Duration) {
+	labels := prometheus.Labels{
+		metrics.HTTPRouteLabel:        route,
+		metrics.HTTPMethodLabel:       method,
+		metrics.HTTPResponseCodeLabel: strconv.Itoa(w.statusCode),
```

Comment:
It's more convenient to use the status code itself because then you can do `sum(rate(http_response_count{code=~"5.."}[5m])) / sum(rate(http_response_count[5m]))` to get the error ratio (matches all 5xx errors). Same w/ "30." for redirects, "2.." for successful responses, etc.

Reply:
ok, sgtm.